nixos/modules/system/boot/systemd/tmpfiles.nix

```diff
@@ -20,6 +20,59 @@ in
       '';
     };
 
+    systemd.tmpfiles.settings = mkOption {
+      description = ''
+        a declarative way to specify tmpfiles rules
+      '';
+      example = {
+        "my-service" = {
+          "/var/lib/my-service/statefolder".d = {
+            mode = "0755";
+            user = "root";
+            group = "root";
+          };
+        };
+      };
+      type = types.attrsOf (types.attrsOf (types.attrsOf (types.submodule ({ config, ... }: {
```

Comment:
I think this is one `attrsOf` too deep. I'd rather write `type = "d";` than use an `attrsOf` for that last level, particularly because it currently indicates you should be able to do this:

```
systemd.tmpfiles.settings.foo."/bar" = {
  d.mode = "0755";
  f.argument = "baz";
};
```

Of course this is nonsense, since one path can only have one entry, but here I've made that path both a directory and a file.

Reply:
> since one path can only have one entry,

you can have multiple entries per path

some types like `f` and `d` conflict but others do not
we need the ability to specify multiple - this is used in nixos and beyond

Reply:
it would be nice to assert those types that can conflict and which not, but I have no good idea how to do that.

Reply:
I stand corrected :)

Reply:
at what point should the check run? I guess if we use writeTextDir we can't really do any verification on the file?

Reply:
That's fair, though I'm not sure systemd actually *has* a validation command for tmpfiles. But if it does and we have to use `runCommand` to validate with it, we should still *definitely* use `passAsFile` rather than shell-escaping arguments to `echo`.

Reply:
There is no validation command for systemd-tmpfiles. I think it's fine to not add a validation derivation here, we can always add a one later.